juju/juju
# Upgrade Series Worker Systemd Service Unit Wrangling

## Description of change

This patch adds a new type to the upgrade-series worker package that is responsible for handling service and agent binary file manipulation based on the to/from OS series.

There is also a new API method for retrieving the target series of the current upgrade, and some miscellaneous drive-by tidy-ups.

The initial implementation conditionally writes systemd service files as required. A forthcoming patch will add agent binary manipulation.

## QA steps

- `export JUJU_DEV_FEATURE_FLAGS=upgrade-series`
- Bootstrap
- Deploy a Trusty charm (Ubuntu or Redis will work)
- `juju upgrade-series prepare 0 xenial`
- SSH on to the machine and check that service files reside under folders in /lib/systemd/system and that there are service symlinks under /etc/systemd/system.

## Documentation changes

None.

## Bug reference

N/A
